firewood/src/merkle/tests/proof.rs

```diff
@@ -0,0 +1,271 @@
+        assert!(!proof.is_empty());
+        proof.verify(key, Some(val), &root_hash).unwrap();
+    }
+}
+
+#[test]
+fn test_proof_end_with_leaf() {
+    let merkle = init_merkle([
+        ("do", "verb"),
+        ("doe", "reindeer"),
+        ("dog", "puppy"),
+        ("doge", "coin"),
+        ("horse", "stallion"),
+        ("ddd", "ok"),
+    ]);
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    let key = b"doe";
+
+    let proof = merkle.prove(key).unwrap();
+    assert!(!proof.is_empty());
+
+    proof.verify(key, Some(b"reindeer"), &root_hash).unwrap();
+}
+
+#[test]
+fn test_proof_end_with_branch() {
+    let items = [
+        ("d", "verb"),
+        ("do", "verb"),
+        ("doe", "reindeer"),
+        ("e", "coin"),
+    ];
+    let merkle = init_merkle(items);
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    let key = b"d";
+
+    let proof = merkle.prove(key).unwrap();
+    assert!(!proof.is_empty());
+
+    proof.verify(key, Some(b"verb"), &root_hash).unwrap();
+}
+
+#[test]
+fn test_bad_proof() {
+    let rng = firewood_storage::SeededRng::from_env_or_random();
+    let set = fixed_and_pseudorandom_data(&rng, 800);
+    let mut items = set.iter().collect::<Vec<_>>();
+    items.sort_unstable();
+    let merkle = init_merkle(items.clone());
+    let root_hash = merkle.nodestore().root_hash().unwrap();
+
+    for (key, value) in items {
+        let proof = merkle.prove(key).unwrap();
+        assert!(!proof.is_empty());
+
+        // Delete an entry from the generated proofs.
+        let mut new_proof = proof.into_mutable();
+        new_proof.pop();
+
+        // TODO: verify error result matches expected error
+        assert!(new_proof.verify(key, Some(value), &root_hash).is_err());
+    }
+}
+
+#[test]
+fn exclusion_with_proof_value_present() {
+    // Build a trie where an ancestor on the path has a value
+    let mut merkle = crate::merkle::tests::create_in_memory_merkle();
+    // Parent has a value
+    merkle.insert(&[0u8], Box::from([0u8])).unwrap();
+    // Child under the same branch
+    merkle.insert(&[0u8, 1u8], Box::from([1u8])).unwrap();
+
+    let merkle = merkle.hash();
+    let root_hash = merkle.nodestore.root_hash().unwrap();
+    debug!("{}", merkle.dump_to_string().unwrap());
+    debug!("root_hash: {root_hash:?}");
+
+    // Non-existent key under the same parent branch
+    let missing = [0u8, 2u8];
+    let proof = merkle.prove(&missing).unwrap();
+
+    debug!("{proof:#?}");
+    assert_eq!(
+        proof.as_ref().first().unwrap().to_hash(),
+        root_hash.clone().into_hash_type()
+    );
+
+    // Ensure at least one node in the proof carries a value (proof value present)
+    assert!(proof.as_ref().iter().any(|n| n.value_digest.is_some()));
+
+    // Exclusion should verify with expected None even if proof includes node values
+    proof
+        .verify(missing, Option::<&[u8]>::None, &root_hash)
+        .unwrap();
```

Comment:
however, fixing that breaks here because I realized just yesterday that `verify` is using the wrong hash fn

https://github.com/ava-labs/firewood/blob/0098d123e103844e2d7dfc032e911be52da80489/storage/src/hashednode.rs#L114-L126

The three other test failures might be because of the same `verify` reason.

Reply:
although, `UnexpectedHash` is only ever produced here so that might not be the reason.

https://github.com/ava-labs/firewood/blob/0098d123e103844e2d7dfc032e911be52da80489/firewood/src/proof.rs#L175-L177

Reply:
Oh, I see. the value is too small to be hashed so we're running into the same issue as above.

I'm not entirely sure how this should be handled. My gut reaction is to make it so that the partialeq/eq impls on HashOrRlp hashes the rlp value of the other value is already hash.